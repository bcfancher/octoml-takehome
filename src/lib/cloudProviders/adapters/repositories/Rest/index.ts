export { RestCloudInstanceTypeRepository } from './RestCloudInstanceTypeRepository'
export { RestCloudProviderRepository } from './RestCloudProviderRepository'
