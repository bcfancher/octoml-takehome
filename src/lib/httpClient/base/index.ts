export { default as iHttpClient } from './iHttpClient'