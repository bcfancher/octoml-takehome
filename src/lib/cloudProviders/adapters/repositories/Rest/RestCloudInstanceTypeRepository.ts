import { inject, injectable } from 'inversify'
import { iCloudInstanceRepository } from '../../../base/repositories'
import { CloudInstanceTypeEntity } from '../../../entities'
import { iHttpClient } from '../../../../httpClient'

const PROVIDERS_BASE_URL = '/api/v1/cloud-instance-types'
const BASE_HEADERS = {
  Accept: 'application/json',
  'Content-Type': 'application/json',
}

@injectable()
export class RestCloudInstanceTypeRepository implements iCloudInstanceRepository {
  constructor(
    @inject(iHttpClient)
    private httpClient: iHttpClient,
  ) {}

  public async getAllInstanceTypes(): Promise<CloudInstanceTypeEntity[]> {
    const allProvidersResponse = await this.httpClient.get({
      url: PROVIDERS_BASE_URL,
      headers: BASE_HEADERS,
    })

    const allProvidersJson = JSON.parse(allProvidersResponse.body ?? '') as CloudInstanceTypeEntity.jsonType[]

    return allProvidersJson.map(
      (provider) => new CloudInstanceTypeEntity(provider)
    )
  }
}
