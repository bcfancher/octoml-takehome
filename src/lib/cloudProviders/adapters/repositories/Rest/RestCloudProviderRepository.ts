import { inject, injectable } from 'inversify'
import { iCloudProviderRepository } from '../../../base/repositories'
import { CloudProviderEntity } from '../../../entities'
import { iHttpClient } from '../../../../httpClient'

const PROVIDERS_BASE_URL = '/api/v1/cloud-providers'
const BASE_HEADERS = {
  Accept: 'application/json',
  'Content-Type': 'application/json',
}

@injectable()
export class RestCloudProviderRepository implements iCloudProviderRepository {
  constructor(
    @inject(iHttpClient)
    private httpClient: iHttpClient,
  ) {}

  public async getAllProviders(): Promise<CloudProviderEntity[]> {
    const allProvidersResponse = await this.httpClient.get({
      url: PROVIDERS_BASE_URL,
      headers: BASE_HEADERS,
    })

    const allProvidersJson = JSON.parse(allProvidersResponse.body ?? '') as CloudProviderEntity.jsonType[]

    return allProvidersJson.map(
      (provider) => new CloudProviderEntity(provider)
    )
  }
}
