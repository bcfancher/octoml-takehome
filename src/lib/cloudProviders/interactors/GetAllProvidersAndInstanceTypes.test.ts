import { Container } from 'inversify'
import { setupServer, SetupServerApi } from 'msw/node'

import { GetAllProvidersAndInstanceTypes } from '.'
import {
  cloudProvidersInteractorsIocModule,
  restCloudProvidersIocModule,
} from '../'
import { fetchHttpClientIocModule } from '../../httpClient'
import { restMswHandler } from '../fixtures/Rest'

let interactor: GetAllProvidersAndInstanceTypes
let mswServer: SetupServerApi

beforeAll(() => {
  mswServer = setupServer(
    ...restMswHandler(),
  )
  mswServer.listen({
    onUnhandledRequest: 'error',
  })

  const container = new Container()
  container.load(
    fetchHttpClientIocModule,
    cloudProvidersInteractorsIocModule,
    restCloudProvidersIocModule,
  )

  interactor = container.get(GetAllProvidersAndInstanceTypes)
})

it('gets collated list of providers and instance types', async () => {
  const actual = await interactor.execute()

  console.log({ actual: actual.map((provider) => provider.toJsonString()) })

  expect(actual).toMatchSnapshot()
})
