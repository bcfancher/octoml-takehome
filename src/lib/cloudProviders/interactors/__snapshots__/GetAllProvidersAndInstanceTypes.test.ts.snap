// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gets collated list of providers and instance types 1`] = `
Array [
  CloudProviderEntity {
    "_id": 1,
    "_instanceTypes": Array [
      CloudInstanceTypeEntity {
        "_id": 6,
        "_key": "m4.large",
        "_name": "m4.large",
        "_numCpus": 2,
        "_provider": [Circular],
        "_providerId": 1,
        "_ramGb": 8,
      },
      CloudInstanceTypeEntity {
        "_id": 7,
        "_key": "m4.xlarge",
        "_name": "m4.xlarge",
        "_numCpus": 4,
        "_provider": [Circular],
        "_providerId": 1,
        "_ramGb": 16,
      },
      CloudInstanceTypeEntity {
        "_id": 8,
        "_key": "m4.2xlarge",
        "_name": "m4.2xlarge",
        "_numCpus": 8,
        "_provider": [Circular],
        "_providerId": 1,
        "_ramGb": 32,
      },
      CloudInstanceTypeEntity {
        "_id": 9,
        "_key": "m4.4xlarge",
        "_name": "m4.4xlarge",
        "_numCpus": 16,
        "_provider": [Circular],
        "_providerId": 1,
        "_ramGb": 64,
      },
      CloudInstanceTypeEntity {
        "_id": 10,
        "_key": "m4.10xlarge",
        "_name": "m4.10xlarge",
        "_numCpus": 40,
        "_provider": [Circular],
        "_providerId": 1,
        "_ramGb": 160,
      },
      CloudInstanceTypeEntity {
        "_id": 11,
        "_key": "m4.16xlarge",
        "_name": "m4.16xlarge",
        "_numCpus": 64,
        "_provider": [Circular],
        "_providerId": 1,
        "_ramGb": 256,
      },
    ],
    "_key": "aws",
    "_name": "Amazon Web Services (AWS)",
  },
  CloudProviderEntity {
    "_id": 2,
    "_instanceTypes": Array [
      CloudInstanceTypeEntity {
        "_id": 1,
        "_key": "c2-standard-4",
        "_name": "c2-standard-4",
        "_numCpus": 4,
        "_provider": [Circular],
        "_providerId": 2,
        "_ramGb": 16,
      },
      CloudInstanceTypeEntity {
        "_id": 2,
        "_key": "c2-standard-8",
        "_name": "c2-standard-8",
        "_numCpus": 8,
        "_provider": [Circular],
        "_providerId": 2,
        "_ramGb": 32,
      },
      CloudInstanceTypeEntity {
        "_id": 3,
        "_key": "c2-standard-16",
        "_name": "c2-standard-16",
        "_numCpus": 16,
        "_provider": [Circular],
        "_providerId": 2,
        "_ramGb": 64,
      },
      CloudInstanceTypeEntity {
        "_id": 4,
        "_key": "c2-standard-30",
        "_name": "c2-standard-30",
        "_numCpus": 30,
        "_provider": [Circular],
        "_providerId": 2,
        "_ramGb": 120,
      },
      CloudInstanceTypeEntity {
        "_id": 5,
        "_key": "c2-standard-60",
        "_name": "c2-standard-60",
        "_numCpus": 60,
        "_provider": [Circular],
        "_providerId": 2,
        "_ramGb": 240,
      },
    ],
    "_key": "gcp",
    "_name": "Google Cloud Platform (GCP)",
  },
]
`;
